---
title: "DESeq2 workflow - Panama SCTLD transmission (O. faveolata host)"
author: "Michael Studivan"
date: 
output:
  html_document:
    theme: darkly
    toc: yes
    toc_depth: 3
    toc_float: yes
knit: (function(input, ...) {
    rmarkdown::render(
      input,
      output_file = 'index.html',
      envir = globalenv()
    )
  })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

<a href="https://github.com/mstudiva/SCTLD-Panama-omics" class="github-corner" aria-label="View source on GitHub"><svg width="80" height="80" viewBox="0 0 250 250" style="fill:#2C3E50; color:#fff; position: absolute; top: 0; border: 0; right: 0;" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a>

```{=html}
<style>.github-corner:hover .octo-arm{animation:octocat-wave 560ms ease-in-out}@keyframes octocat-wave{0%,100%{transform:rotate(0)}20%,60%{transform:rotate(-25deg)}40%,80%{transform:rotate(10deg)}}@media (max-width:500px){.github-corner:hover .octo-arm{animation:none}.github-corner .octo-arm{animation:octocat-wave 560ms ease-in-out}}</style>
```

#### version: `r library(magrittr)` `r Sys.Date() %>% format(format="%d %B, %Y")`

#### [GitHub repository](https://github.com/mstudiva/SCTLD-Panama-omics){target="_blank"}

## A B O U T &nbsp; T H I S &nbsp; D O C U M E N T

This walkthrough describes the differential expression analysis of *Orbicella faveolata* gene counts from a stony coral tissue loss disease experiment. It covers (1) dataframe setup and outlier detection, (2) data visualization and statistical testing, (3) differential expression analysis, and (4) p value export for downstream GO and KOG analyses. For initial processing of TagSeq reads regardless of species, please see below: 

- [TagSeq processing](https://mstudiva.github.io/SCTLD-Panama-omics/code/)

Library prep, bioinformatics, and analysis protocols are credited to the TagSeq pipeline originally developed by Eli Meyer and Misha Matz: https://doi.org/10.1111/j.1365-294X.2011.05205.x, and further refined by Michael Studivan: https://github.com/mstudiva/tag-based_RNAseq

<br>

```{r setup, include=FALSE}
# ---- user-configurable paths ----
counts_file  <- "data/allcounts_ofav.txt"
design_file  <- "data/design.csv"
annotation_file <- "annot/Ofaveolata_iso2geneName.tab"  # optional (for cherry-picking)

# ---- analysis parameters ----
min_row_sum_counts <- 10
wgcna_min_count <- 10
wgcna_max_prop_low <- 0.90
outs <- c(0)     # outlier sample indices to remove (set to integer(0) if none)
alpha <- 0.10     # DESeq2 adjusted p-value threshold
```

## Packages

```{r packages}
suppressPackageStartupMessages({
  library(DESeq2)
  library(BiocParallel)
  library(arrayQualityMetrics)
  library(Biobase)
  library(pheatmap)
  library(vegan)
  library(ape)
  library(VennDiagram)
  library(dplyr)
  library(stringr)
  library(ggpubr)
})
```

---

## Data import

```{r data-import}
counts <- read.table(counts_file)

# removing the parent sample (column 41 in original script)
if (ncol(counts) >= 41) {
  counts <- subset(counts, select = -c(41))
}

keep <- rowSums(counts) >= min_row_sum_counts
countData <- counts[keep, ]

dir.create("outputs", showWarnings = FALSE)
write.csv(countData, file = "outputs/countData.csv")

# WGCNA-style filter: remove genes with counts < wgcna_min_count in >= wgcna_max_prop_low of samples
counts4wgcna <- counts[apply(counts, 1, function(x) sum(x < wgcna_min_count)) < ncol(counts) * wgcna_max_prop_low, ]
write.csv(counts4wgcna, file = "outputs/counts4wgcna.csv")

design <- read.csv(design_file, head = TRUE)
design <- design[design$genotype != "Parent", ]
stopifnot(nrow(design) == ncol(countData))
str(design)
```

---

## Model design and outlier QC

```{r model-and-qc}
dds <- DESeqDataSetFromMatrix(countData = countData, colData = design, design = ~ genotype + fate)

# set factor levels for fate
dds$fate <- factor(dds$fate, levels = c("healthy", "nai", "diseased"))

Vsd <- varianceStabilizingTransformation(dds)

e <- ExpressionSet(assay(Vsd), AnnotatedDataFrame(as.data.frame(colData(Vsd))))

# arrayQualityMetrics creates an HTML report directory
# If you knit on GitHub Actions, consider setting `force = TRUE` and `do.logtransform = FALSE`.
arrayQualityMetrics(e, intgroup = c("fate"), force = TRUE)

# remove outliers by index (if any)
if (length(outs) > 0) {
  countData <- countData[, -outs, drop = FALSE]
  Vsd <- Vsd[, -outs]
  counts4wgcna <- counts4wgcna[, -outs, drop = FALSE]
  design <- design[-outs, , drop = FALSE]
}

# remake model after outlier removal
dds <- DESeqDataSetFromMatrix(countData = countData, colData = design, design = ~ genotype + fate)
dds$fate <- factor(dds$fate, levels = c("healthy", "nai", "diseased"))

save(dds, design, countData, Vsd, counts4wgcna, file = "outputs/initial.RData")

# variance-stabilized matrix for ordination / heatmaps
vsd <- assay(Vsd)
snames <- paste(colnames(countData), design[, 4], design[, 6], sep = ".")
colnames(vsd) <- snames
save(vsd, design, file = "outputs/vsd.RData")

# WGCNA-ready VST (optional)
wg <- DESeqDataSetFromMatrix(countData = counts4wgcna, colData = design, design = ~ genotype + fate)
vsd.wg <- assay(varianceStabilizingTransformation(wg), blind = TRUE)
colnames(vsd.wg) <- snames
save(vsd.wg, design, file = "outputs/data4wgcna.RData")
```

---

## Heatmap, PCoA, PERMANOVA

```{r ordination}
load("outputs/vsd.RData")

# sample correlation heatmap
pheatmap(cor(vsd))

conditions <- design
conditions$fate <- factor(conditions$fate, levels = c("healthy", "nai", "diseased"))

# PCoA (Manhattan distance; scaled to match original script)
dds.pcoa <- pcoa(dist(t(vsd), method = "manhattan") / 1000)
scores <- dds.pcoa$vectors

# eigenvalues (% variation explained in Relative_eig)
dds.pcoa$values
plot(dds.pcoa$values$Relative_eig, ylab = "Relative eigenvalue", xlab = "PCoA axis")
points(dds.pcoa$values$Broken_stick, col = "red", pch = 3)

# plot by fate and treatment (if treatment column exists)
par(mfrow = c(1, 2))
plot(scores[, 1], scores[, 2],
     col = c("green", "orange", "red")[as.numeric(as.factor(conditions$fate))],
     xlab = "Coordinate 1", ylab = "Coordinate 2", main = "Fate")
ordispider(scores, conditions$fate, label = FALSE, col = c("green", "orange", "red"))
legend("topright", legend = c("healthy", "NAI", "diseased"), fill = c("green", "orange", "red"), bty = "n")

if ("treatment" %in% names(conditions)) {
  plot(scores[, 1], scores[, 2],
       col = c("green", "black", "red")[as.numeric(as.factor(conditions$treatment))],
       xlab = "Coordinate 1", ylab = "Coordinate 2", main = "Treatment")
  ordispider(scores, conditions$treatment, label = FALSE, col = c("green", "black", "red"))
  legend("topleft", legend = levels(as.factor(conditions$treatment)), fill = c("green", "red"), bty = "n")
}

par(mfrow = c(1, 1))

# neighbor-joining tree using first 4 axes
tre <- nj(dist(scores[, 1:4]))
plot(tre, cex = 0.8)

# PERMANOVA
ad <- adonis(t(vsd) ~ genotype + fate, data = conditions, method = "manhattan", permutations = 1e6)
ad
```

---

## DESeq2 models

```{r deseq-models}
load("outputs/initial.RData")

# full model (for contrasts)
dds <- DESeq(dds, parallel = TRUE)

# LRT for multi-level fate effect
dds$fate <- factor(dds$fate, levels = c("healthy", "nai", "diseased"))
dds_fate <- DESeq(dds, test = "LRT", reduced = ~ genotype, parallel = TRUE)

save(dds, dds_fate, file = "outputs/realModels.RData")
```

---

## DEGs and contrasts

```{r degs}
load("outputs/realModels.RData")

# main effects
fate <- results(dds_fate)
genotype <- results(dds)

degs_fate <- rownames(fate)[fate$padj < alpha & !is.na(fate$padj)]
degs_genotype <- rownames(genotype)[genotype$padj < alpha & !is.na(genotype$padj)]

# pairwise contrasts for fate
diseased_healthy <- results(dds, contrast = c("fate", "diseased", "healthy"))
nai_healthy      <- results(dds, contrast = c("fate", "nai", "healthy"))
diseased_nai     <- results(dds, contrast = c("fate", "diseased", "nai"))

degs_diseased_healthy <- rownames(diseased_healthy)[diseased_healthy$padj < alpha & !is.na(diseased_healthy$padj)]
degs_nai_healthy      <- rownames(nai_healthy)[nai_healthy$padj < alpha & !is.na(nai_healthy$padj)]
degs_diseased_nai     <- rownames(diseased_nai)[diseased_nai$padj < alpha & !is.na(diseased_nai$padj)]

save(fate, genotype, diseased_healthy, nai_healthy, diseased_nai, file = "outputs/pvals.RData")

# DEG abundance density
load("outputs/vsd.RData")
means <- apply(vsd, 1, mean)

plot(density(means), main = "Mean VST abundance density", xlab = "Mean VST")
lines(density(means[degs_genotype]), col = "blue")
lines(density(means[degs_fate]), col = "orange")
legend("topright", title = "Factor", legend = c("genotype", "fate"), fill = c("blue", "orange"))
```

---

## Venn diagrams

```{r venn}
load("outputs/pvals.RData")

candidates <- list("genotype" = degs_genotype, "fate" = degs_fate)

fullmodel_venn <- venn.diagram(
  x = candidates,
  filename = NULL,
  col = "transparent",
  fill = c("blue", "orange"),
  alpha = 0.5,
  cex = 1.5,
  fontfamily = "sans"
)
grid::grid.newpage()
grid::grid.draw(fullmodel_venn)

pairwise <- list(
  "diseased_healthy" = degs_diseased_healthy,
  "nai_healthy"      = degs_nai_healthy,
  "diseased_nai"     = degs_diseased_nai
)

pairwise_venn <- venn.diagram(
  x = pairwise,
  filename = NULL,
  col = "transparent",
  fill = c("blue", "orange", "lightblue"),
  alpha = 0.5,
  cex = 1.2,
  fontfamily = "sans"
)
grid::grid.newpage()
grid::grid.draw(pairwise_venn)
```

---

## GO/KOG export tables

```{r go-kog}
load("outputs/pvals.RData")

# signed log p-values helper
signed_logp <- function(res) {
  source <- res[!is.na(res$pvalue), ]
  out <- data.frame(gene = rownames(source))
  out$lpv <- -log10(source$pvalue)
  out$lpv[source$stat < 0] <- out$lpv[source$stat < 0] * -1
  out
}

genotype.p <- signed_logp(genotype)
fate.p <- signed_logp(fate)

write.csv(genotype.p, file = "outputs/genotype_lpv.csv", row.names = FALSE, quote = FALSE)
write.csv(fate.p,     file = "outputs/fate_lpv.csv",     row.names = FALSE, quote = FALSE)

# fold-change tables for pairwise contrasts
fc_table <- function(res) {
  source <- res[!is.na(res$pvalue), ]
  data.frame(gene = rownames(source), lfc = source$log2FoldChange)
}

write.csv(fc_table(diseased_healthy), file = "outputs/diseased_healthy_fc.csv", row.names = FALSE, quote = FALSE)
write.csv(fc_table(nai_healthy),      file = "outputs/nai_healthy_fc.csv",      row.names = FALSE, quote = FALSE)
write.csv(fc_table(diseased_nai),     file = "outputs/diseased_nai_fc.csv",     row.names = FALSE, quote = FALSE)

write.csv(signed_logp(diseased_healthy), file = "outputs/diseased_healthy_lpv.csv", row.names = FALSE, quote = FALSE)
write.csv(signed_logp(nai_healthy),      file = "outputs/nai_healthy_lpv.csv",      row.names = FALSE, quote = FALSE)
write.csv(signed_logp(diseased_nai),     file = "outputs/diseased_nai_lpv.csv",     row.names = FALSE, quote = FALSE)
```

---

## Optional: immune “cherry-picking” by annotation keywords

```{r cherrypick, eval=FALSE}
# Set eval=TRUE if you have the annotation_file present in data/
load("outputs/pvals.RData")

annot <- read.table(annotation_file, sep = "\t", quote = "", fill = FALSE) %>%
  mutate(gene = V1, annot = V2) %>%
  dplyr::select(gene, annot)

cherrypicking <- diseased_healthy.p %>%
  dplyr::filter(abs(lpv) >= 1) %>%
  dplyr::left_join(annot, by = "gene") %>%
  dplyr::filter(str_detect(
    annot,
    "NF-kappaB|peroxidas|TGF-beta|protein tyrosine kinase|fibrinogen|WD repeat-containing protein|apoptosis|extracellular matrix"
  ))

write.csv(cherrypicking, file = "outputs/trans_mcav_cherrypicking.csv", row.names = FALSE)
```

---

## Gene-level exports (plotCounts)

```{r gene-counts}
load("outputs/realModels.RData")

genes_of_interest <- c(
  "Mcavernosa14879","Mcavernosa98966","Mcavernosa64646","Mcavernosa69648",
  "Mcavernosa9810","Mcavernosa43816","Mcavernosa54510","Mcavernosa47647",
  "Mcavernosa50735","Mcavernosa61972","Mcavernosa14843","Mcavernosa16729",
  "Mcavernosa184695","Mcavernosa49548","Mcavernosa59808","Mcavernosa9650",
  "Mcavernosa102943","Mcavernosa12949","Mcavernosa29964","Mcavernosa20827",
  "Mcavernosa71973","Mcavernosa126229","Mcavernosa27667","Mcavernosa21718",
  "Mcavernosa96261","Mcavernosa366959"
)

dir.create("outputs/gene_counts", showWarnings = FALSE)

for (g in genes_of_interest) {
  df <- plotCounts(dds, gene = g, intgroup = "fate", returnData = TRUE)
  write.csv(df, file = file.path("outputs/gene_counts", paste0(g, ".csv")), row.names = FALSE)
}

head(read.csv(file.path("outputs/gene_counts", paste0(genes_of_interest[1], ".csv"))))
```

---

## Session info

```{r session-info}
sessionInfo()
```
